### PR TITLE
feat: Add `.parserConfiguration()` method, deprecating package.json config

### PR DESCRIPTION
**Pull request by jean-emmanuel**
A non-breaking proposal to address #1259 and #1248, it might be incomplete so please amend if needed.

### CODE REVIEW

**Comment by bcoe** [comment]
@jean-emmanuel this looks great, we should add a test though 👍

**Comment by jean-emmanuel** [comment]
I just pushed 2 tests, what do you think ?

**Comment by jean-emmanuel** [comment]
Travis' fail seems unrelated by the way.

File: yargs.js [hunk 1]
```diff
@@ -1010,7 +1018,14 @@ function Yargs (processArgs, cwd, parentRequire) {
     args = args || processArgs
 
     options.__ = y18n.__
-    options.configuration = pkgUp()['yargs'] || {}
+    options.configuration = self.getParserConfiguration()
+
+    // Deprecated
+    let pkgConfig = pkgUp()['yargs']
+    if (pkgConfig) {
+      console.warn('Configuring yargs through package.json is deprecated and will be removed in the next major release, please use the JS API instead.')
```
Comment (evocateur) [inline yargs.js]: There's no other "raw" `console.*` call in this file, we should probably use the instance-scoped [`_logger.error()`](https://github.com/yargs/yargs/blob/27bf73923423dbe84dd2fd282fdd31d26bdb6cee/yargs.js#L956-L975) here instead.

Parenthetically, `console.warn` vs `console.error` is just semantics, it doesn't call `process.exit()` and both console methods log to `process.stderr` anyway.

Comment (jean-emmanuel) [reply yargs.js]: Alright, thanks for reviewing the PR ! 41fe766

Comment (jean-emmanuel) [reply yargs.js]: Using the scoped logger happens to break the tests as this deprecation warning is unexpected. I've reverted the commit, I'm not sure it's relevant to pass this warning for non-cli print uses.

**Comment by bcoe** [comment]
@evocateur @jean-emmanuel where does this stand? would love to land the work; I'm thinking perhaps we land this as a major, and then in the next major actually retire it?

File: yargs.js (hunk 1 above)
Comment (evocateur) [reply yargs.js]: Okay, fair enough. A pity about the brittle tests.

**Comment by evocateur** [comment]
This strikes me as semver-minor, with removal in the next major. If there are other semver-majors waiting in the wings, it could certainly go along with that.

**Comment by bcoe** [comment]
@evocateur @jean-emmanuel, @isaacs just pointed this out to me:

https://nodejs.org/api/util.html#util_util_deprecate_fn_msg_code

☝️ Node.js has a built in deprecate message that debounces.

**Comment by bcoe** [comment]
let's use this -- @evocateur there's some somewhat finicky breaking changes landing in `yargs-parser`, so I'm a bit tempted to land it in a major (the ones you reviewed around handling of missing right-hand values).

File: yargs.js [hunk 2]
```diff
@@ -1010,7 +1018,14 @@ function Yargs (processArgs, cwd, parentRequire) {
     args = args || processArgs
 
     options.__ = y18n.__
-    options.configuration = pkgUp()['yargs'] || {}
+    options.configuration = self.getParserConfiguration()
+
+    // Deprecated
+    let pkgConfig = pkgUp()['yargs']
```
Comment (evocateur) [inline yargs.js]: I like @bcoe's suggestion to use [`util.deprecate()`](https://nodejs.org/api/util.html#util_util_deprecate_fn_msg_code) here. It'll require a minor extraction of the logic, but that's arguably a good thing anyway.

```js
const util = require('util');

const getPkgConfig = util.deprecate(() => pkgUp()['yargs'], 'Configuring yargs through package.json is deprecated...');

// etc etc
```

We can test it by listening to the [warning](https://nodejs.org/api/process.html#process_event_warning) event on `process`:
```js
let called = false
process.once('warning', () => {
  called = true
})
// do the thing
// assert called === true ?
```

Comment (jean-emmanuel) [reply yargs.js]: Wouldn't this emit the warning whether package.json contains yargs configuration tokens or not ? This deprecates the function that checks for package.json configuration, I think it makes more sense to only warn when the deprecated feature is actually used.

Comment (evocateur) [reply yargs.js]: Oh, yeah, you're right, nevermind `util.deprecate`

**Comment by evocateur** [comment]
@bcoe As @jean-emmanuel rightly pointed out, `util.deprecate()` isn't appropriate for something that is called _every_ time yargs parses arguments.

**Comment by evocateur** [comment]
Thanks @jean-emmanuel!

**Comment by jean-emmanuel** [comment]
woohooo !

**Comment by bcoe** [comment]
this is now available in `yargs@13.1.0`.

**Comment by plroebuck** [comment]
Uh, yeah.
* yargs-13.0.0 mentions in ["breaking changes"](https://github.com/yargs/yargs/blob/master/CHANGELOG.md#1300-2019-02-02) that Node-6 was being dropped (though "package.json" still shows `"node": ">=6"`). **Did you or didn't you?**
* Mocha _will_ support Node-6 _at least_ until end-of-life, so cannot upgrade to yargs-13 yet if support dropped (or will be without another semver-major).
* Yargs-13 introduced `yargs.parserConfiguration`, but no backport to yargs-12 available.
* Users using yargs-13 now get deprecation warning, with no visible means to address the problem from Mocha via yargs-12.

Recommendations?

**Comment by plroebuck** [comment]
BTW, your documentation shows:
```js
yargs.parserConfiguration({
  "yargs": {
    "short-option-groups": true,
    "camel-case-expansion": true,
    "dot-notation": true,
    "parse-numbers": true,
    "boolean-negation": true
  }
})
```
but test fixture ["test/fixtures/configured-bin.js"](https://github.com/yargs/yargs/blob/3c6869aae7b488d2416f66c19a801c06243c075c/test/fixtures/configured-bin.js#L3) shows:
```js
var argv = require('./yargs/index.js')
  .parserConfiguration({'dot-notation': true})
  .help('help')
  .version()
  .argv
```

**Comment by bcoe** [comment]
> * yargs-13.0.0 mentions in ["breaking changes"](https://github.com/yargs/yargs/blob/master/CHANGELOG.md#1300-2019-02-02) that Node-6 was being dropped (though "package.json" still shows `"node": ">=6"`). **Did you or didn't you?**

@plroebuck we can no longer promise that `Node >=6` will continue to be supported in yargs 13; one reason being flakiness we were seeing in CI/CD with the current combination of npm and Node.js shipped on 6.x. The engine field was just missed being updated.

> * Mocha _will_ support Node-6 _at least_ until end-of-life, so cannot upgrade to yargs-13 yet if support dropped (or will be without another semver-major).

We would be open to back-porting this change-request, and security patches back to `12.x`.

> BTW, your documentation shows...

☝️ I believe the documentation is wrong.

@plroebuck could you open a few issues rather than commenting on a closed issue, the odds of us losing track of your comment are high.

@jean-emmanuel interested in back-porting this to `yargs@12.x` without the deprecation warning. To do so we would create a new `12.x` branch, from the point at which we released https://github.com/yargs/yargs/commit/8789bf4f56940248316c58997b107b029dbdb297, and cherry-pick create a pull request against this branch.

**Pull request merged**